### PR TITLE
@grafana/runtime: User storage

### LINKED ISSUE #91664: Add ability to save any kind of data per user
**Why is this needed**:

Sometimes panels should save some data per user. Like favorited items, settings, etc. Currently, one of possible solution is to save the data in `localStorage`.  But it's limited by using the same browser for user to get access to his saved data. The data should be available from any browser/device for the user.

**What would you like to be added**:

Easy way to save/read any kind of data for the current user. Probably, it will be stored as a json column in the same table with user preferences (timezone, theme, etc.)

Example of usage:

```
import { userStorage } from '@grafana/runtime';

const data = userStorage.get('key');

userStorage.set('key', [1, 10]);
```

**Who is this feature for?**

Plugin developers

### PR DESCRIPTION
**Pull request by andresmgot**
**What is this feature?**

Requires https://github.com/grafana/grafana/pull/95557

Add `getItem` and `setItem` functions to use the new backend storage for storing user information. In case that's not available (if the feature flag is not available or the request fails), it falls back to use the browser `localStorage`.

See how could be used in this example https://github.com/grafana/grafana-plugin-examples/pull/414

**Why do we need this feature?**

Plugins and other services want to store settings per user and be able to persist that information among different browsers.

**Who is this feature for?**

Devs

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/91664

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: packages/grafana-runtime/src/utils/userStorage.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      // Fallback to localStorage
+      return localStorage.getItem(resourceName);
+    }
+    const userStorage = await getUserStorage(resourceName);
+    if (!userStorage) {
+      // Also, fallback to localStorage for backward compatibility once userStorageAPI is enabled
+      return localStorage.getItem(resourceName);
+    }
+    return userStorage.spec.data[key];
+  }
```
Comment (joshhunt) [inline packages/grafana-runtime/src/utils/userStorage.ts:91]: Maybe using IndexedDB might be a closer match to the semantics of a backend API anyway? That would avoid mixing async and sync 'backends' as well?

File: packages/grafana-runtime/src/utils/userStorage.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+   */
+  async setItem(key: string, value: string): Promise<void> {
+    const resourceName = getResourceName(this.service);
+    if (!canUseUserStorage()) {
+      // Fallback to localStorage
+      localStorage.setItem(key, value);
+      return;
+    }
+
+    const userStorage = await getUserStorage(resourceName);
```
Comment (joshhunt) [inline packages/grafana-runtime/src/utils/userStorage.ts:107]: kinda annoying that setting a value takes two API calls.

File: packages/grafana-runtime/src/utils/userStorage.ts (hunk 1 above)
Comment (andresmgot) [reply packages/grafana-runtime/src/utils/userStorage.ts:91]: currently, people are already using the `localStorage` for this so I prefer to keep this simple interface so the replacement is as easy as posisble (the only change is the switch from sync to async).

File: packages/grafana-runtime/src/utils/userStorage.ts (hunk 2 above)
Comment (andresmgot) [reply packages/grafana-runtime/src/utils/userStorage.ts:107]: very good point, I have changed the class so the state is stored in it (so only one GET request is ever made).

I have also modified the PUT request with a PATCH one so we don't need to send the whole object.

**Comment by andresmgot** [comment]
Thanks for the review @joshhunt @ashharrison90! Do you mind taking a look again?

**Review by joshhunt — COMMENTED** [review]
Looks good 👍 

How do we expect consumers to use this? Each would create their own instance of this class?

It looks like on the backend the storage is scoped to the plugin (the service prop) - should local storage keys be scoped the same?
It might be unexpected that _sometimes_ the keys exist in a global scope and sometimes they don't.

I'll let one of your Plugins Platformer frontend devs handle final approval for it.

File: packages/grafana-runtime/src/utils/userStorage.ts (hunk 1 above)
Comment (joshhunt) [reply packages/grafana-runtime/src/utils/userStorage.ts]: I think indexeddb is the better local match for this, but not strongly enough.

**Comment by andresmgot** [comment]
> Looks good 👍

Thanks! 

> How do we expect consumers to use this? Each would create their own instance of this class?

Yes, I have a working example here if you want to take a look: https://github.com/grafana/grafana-plugin-examples/pull/414 

> It looks like on the backend the storage is scoped to the plugin (the service prop) - should local storage keys be scoped the same? It might be unexpected that _sometimes_ the keys exist in a global scope and sometimes they don't.

Not sure if I understand what you mean here, regardless of the storage method (backend or local), when using this library, the service (or plugin ID) will be always part of the entry name and will be scoped. When using the backend, the resource can be retrieved with `GET /apis/userstorage.grafana.app/.../<service>:<user_uid>` and when using the local storage it will be `localStorage.get("<service>:<user_uid>")`.

File: packages/grafana-runtime/src/utils/userStorage.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  private service: string;
+  private resourceName: string;
+  private userUID: string;
+  private canUseUserStorage: boolean;
+  private storageSpec: UserStorageSpec | null | undefined;
+
+  constructor(service: string) {
+    this.service = service;
+    this.userUID = config.bootData.user.uid === '' ? config.bootData.user.id.toString() : config.bootData.user.uid;
+    this.resourceName = `${service}:${this.userUID}`;
```
Comment (leventebalogh) [inline packages/grafana-runtime/src/utils/userStorage.ts:49]: Probably a silly question, but I guess the user UID postfix is validated on the backend, too?

**Review by leventebalogh — APPROVED** [review]
LGTM 👍 

More like a question: did we think about making it also a per-plugin storage, e.g. plugins could only write to specs that are prefixed with their id, so they cannot override each others data?

**Comment by sunker** [comment]
> More like a question: did we think about making it also a per-plugin storage, e.g. plugins could only write to specs that are prefixed with their id, so they cannot override each others data?

Yeah not sure how big of a problem this is, but would prefer to only give plugins access to an already scoped user storage so they can only read/write user data prefixed with its own plugin id. 

Can we instead wrap plugin components in a user storage context providers and add a hook that gives them access to `getItem` and `setItem` only?

File: packages/grafana-runtime/src/utils/userStorage.ts (hunk 3 above)
Comment (andresmgot) [reply packages/grafana-runtime/src/utils/userStorage.ts:49]: Correct, the backend checks that the user UID set in the name of the request and the one in the authentication info matches, if not, it returns a 403.

**Comment by andresmgot** [comment]
Thanks for the feedback @leventebalogh @sunker !

> More like a question: did we think about making it also a per-plugin storage, e.g. plugins could only write to specs that are prefixed with their id, so they cannot override each others data?

That's the idea of the `service` property of the `constructor`. See how it's used [here](https://github.com/grafana/grafana-plugin-examples/pull/414/files#diff-633e497e34c3e3f4b454a01042f915abda7b3e69db754c9a5a7d2179fa7fc3d4R8). The idea is that plugins instantiate a user storage with their plugin ID as the service identifier so they don't overwrite others data.
 
> Yeah not sure how big of a problem this is, but would prefer to only give plugins access to an already scoped user storage so they can only read/write user data prefixed with its own plugin id.

Same, the idea here is to use the service identifier to avoid collisions. Note that the goal is that this could be used for other things that are not plugins (e.g. Grafana core features) so we cannot assume that `service == plugin ID` and that there will always be a plugin ID.

We could expose some kind of utility function specific for plugins, something like `getUserStorageForPlugin(...)` but as far as I know, it would not be possible to neither automatically get a plugin ID from `@grafana/runtime` nor verify that the plugin ID passed is the correct one. Please correct me if I am wrong!

> Can we instead wrap plugin components in a user storage context providers and add a hook that gives them access to `getItem` and `setItem` only?

Similar to the above, I think it will be useful to not limit this to components (so it's a real replacement for `localStorage`) so people can still use it everywhere. Again, if you are proposing something that is not React-component specific and I am missing it let me know!

**Comment by sunker** [comment]
> That's the idea of the `service` property of the `constructor`. See how it's used [here](https://github.com/grafana/grafana-plugin-examples/pull/414/files#diff-633e497e34c3e3f4b454a01042f915abda7b3e69db754c9a5a7d2179fa7fc3d4R8). The idea is that plugins instantiate a user storage with their plugin ID as the service identifier so they don't overwrite others data.

But plugin authors could provide the service name of another plugin and read/overwrite its user data? 

All plugin components _should_ be wrapped in a [PluginContextProvider](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/context/plugins/PluginContextProvider.tsx/#L11-L14). Can you make `userStorage` a prop in the `PluginContextProviderProps` type and instantiate the UserStorage class with the plugin id defined in the plugin meta? Then you can add a `useUserStorage` hook that plugin authors can use to set and get items. 

Then perhaps the UserStorage class can be moved to `public/app/features`. I'm not too concerned about miss-use among core features so they can instantiate the storage with a service name by choice.

**Comment by andresmgot** [comment]
> But plugin authors could provide the service name of another plugin and read/overwrite its user data?

Yes, note that there is no validation in the backend to validate the _service_ calling the user storage, only the user UID is verified (to avoid one user accessing other user storage). It's not the goal to provide service isolation (just the same way that you don't have that using `localStorage`). It's also obviously not the goal to use this to store secrets since the data is not encrypted either.

To clarify, this PR is just an utility helper around an HTTP request so regardless of how we implement it here, all the user storage info is accessible to all the services (as long as it knows the service identifier).

> All plugin components should be wrapped in a [PluginContextProvider](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/context/plugins/PluginContextProvider.tsx/#L11-L14). Can you make userStorage a prop in the PluginContextProviderProps type and instantiate the UserStorage class with the plugin id defined in the plugin meta? Then you can add a useUserStorage hook that plugin authors can use to set and get items.

I think this would unnecesarily limit the feature to Components. As it is right now, if people want to use storage in their Datasource class they can, but if we use the PluginContextProvider, they would not.

> Then perhaps the UserStorage class can be moved to `public/app/features`. I'm not too concerned about miss-use among core features so they can instantiate the storage with a service name by choice.

As far as I can see, code from `@grafana/runtime` cannot import code from `public/app/features`, am I wrong?

**Comment by andresmgot** [comment]
Ping @sunker

**Comment by andresmgot** [comment]
> > I think this would unnecesarily limit the feature to Components. As it is right now, if people want to use storage in their Datasource class they can, but if we use the PluginContextProvider, they would not.
> 
> For data sources we could attach a scoped instance of the user storage class to the datasource instance. So in the DatasourceSrc.loadDatasource you add something along those lines:
> 
> ```ts
> if (instance instanceof DataSourceWithBackend) {
>   instance.userStorage = new UserStorage(instanceSettings.meta.id);
> }
> ```

Sorry, I am not sure what you are referring to, you mean this `loadDatasource`? https://github.com/grafana/grafana/blob/sandboxPluginLoaderGrafana/public/app/features/plugins/datasource_srv.ts#L147

I am not familiar with this part of code so I am not really confident modifying that just for this feature but in any case, note that we'd need that for all plugins, not only datasources (with or without a backend).

> > As far as I can see, code from @grafana/runtime cannot import code from public/app/features, am I wrong?
> 
> Maybe missing something, but why does the user storage class have to be in runtime? Can't we move it to for example `public/app/features/userStorage`?

The class doesn't have to be in @grafana/runtime, we only need to expose something so plugins can do it.

**Review by sunker — APPROVED** [review]
Great stuff! 🚀  Left some nits and minor suggestions. Feel free to ignore if you want.

**Pull request merged**